src/Downloader.ts

```diff
@@ -169,52 +176,69 @@ class Downloader {
 
       //* Objects order in array matters!
-      this.baseUrlForMainPage = basicURLDirector.buildDownloaderBaseUrl([
-        { condition: await MediaWiki.hasWikimediaDesktopApi(), value: MediaWiki.WikimediaDesktopApiUrl.href },
-        { condition: await MediaWiki.hasVisualEditorApi(), value: MediaWiki.visualEditorApiUrl.href },
-        { condition: await MediaWiki.hasWikimediaMobileApi(), value: MediaWiki.WikimediaMobileApiUrl.href },
-      ])
+      const mainPageCapabilitiesList = [
+        { condition: await MediaWiki.hasWikimediaDesktopApi(), value: MediaWiki.WikimediaDesktopApiUrl.href, Director: WikimediaDesktopURLDirector },
+        { condition: await MediaWiki.hasVisualEditorApi(), value: MediaWiki.VisualEditorApiUrl.href, Director: VisualEditorURLDirector },
+        { condition: await MediaWiki.hasWikimediaMobileApi(), value: MediaWiki.WikimediaMobileApiUrl.href, Director: WikimediaMobileURLDirector },
+      ]
+      this.baseUrlForMainPage = basicURLDirector.buildDownloaderBaseUrl(mainPageCapabilitiesList)
+      this.mainPageUrlDirector = this.getUrlDirector(mainPageCapabilitiesList)
     } else {
       switch (forceRender) {
         case 'WikimediaDesktop':
           if (MediaWiki.hasWikimediaDesktopApi()) {
             this.baseUrl = MediaWiki.WikimediaDesktopApiUrl.href
             this.baseUrlForMainPage = MediaWiki.WikimediaDesktopApiUrl.href
+            this.articleUrlDirector = this.mainPageUrlDirector = new WikimediaDesktopURLDirector(MediaWiki.WikimediaDesktopApiUrl.href)
             break
           }
           break
         case 'VisualEditor':
           if (MediaWiki.hasVisualEditorApi()) {
-            this.baseUrl = MediaWiki.visualEditorApiUrl.href
-            this.baseUrlForMainPage = MediaWiki.visualEditorApiUrl.href
+            this.baseUrl = MediaWiki.VisualEditorApiUrl.href
+            this.baseUrlForMainPage = MediaWiki.VisualEditorApiUrl.href
+            this.articleUrlDirector = this.mainPageUrlDirector = new VisualEditorURLDirector(MediaWiki.VisualEditorApiUrl.href)
             break
           }
           break
         case 'WikimediaMobile':
           if (MediaWiki.hasWikimediaMobileApi()) {
             this.baseUrl = MediaWiki.WikimediaMobileApiUrl.href
             this.baseUrlForMainPage = MediaWiki.WikimediaMobileApiUrl.href
+            this.articleUrlDirector = this.mainPageUrlDirector = new WikimediaMobileURLDirector(MediaWiki.WikimediaMobileApiUrl.href)
             break
           }
           break
         default:
           throw new Error('Unable to find specific API end-point to retrieve article HTML')
       }
     }
-    logger.log('Base Url: ', this.baseUrl)
-    logger.log('Base Url for Main Page: ', this.baseUrlForMainPage)
+  }
 
-    if (!this.baseUrl || !this.baseUrlForMainPage) throw new Error('Unable to find appropriate API end-point to retrieve article HTML')
+  public getArticleUrl(dump: Dump, articleId: string): string {
```

Comment:
Don't take `dump` as parameter, only the `articleId`. Create in addition a `getMainPageUrl(articleId)`... the function interface is not great I know.

Reply:
Created **downloader.getMainPageUrl()**